### PR TITLE
Impl ExpandDims op and fallback to CPU if needed

### PR DESCRIPTION
**Pull request by gujinghui**
Impl ExpandDims op and fallback to CPU if needed

### CODE REVIEW

File: caffe2/ideep/operators/expand_squeeze_dims_op.cc [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+    dims_.erase(std::unique(dims_.begin(), dims_.end()), dims_.end());
+    if (dims_.size() < originalSize) {
+      LOG(WARNING) << "Parameter `dims` has repeated dimensions.";
+    }
+    CAFFE_ENFORCE(dims_.front() >= 0, "Dimension ids must be non-negative.");
+  }
+  virtual ~IDEEPSqueezeOp() {}
+
+  bool RunOnDevice() override {
+    if (!OperatorBase::InputBlob(INPUT).template IsType<itensor>()) {
```
Comment (yinghai) [inline caffe2/ideep/operators/expand_squeeze_dims_op.cc]: Have we tested this condition in unit test?

**Comment by gujinghui** [comment]
depends on #15726

@yinghai 
rebased and added unit case.
pls help review.

**Comment by gujinghui** [comment]
@yinghai 
rebased. pls help review.

**Comment by gujinghui** [comment]
@yinghai 

The failure should be not caused by this PR.

> Jan 11 07:47:00 Downloading: "https://download.pytorch.org/models/resnet18-5c106cde.pth" to /var/lib/jenkins/.torch/models/resnet18-5c106cde.pth
Jan 11 07:47:00 ERROR
Jan 11 07:47:00 test_check_onnx_broadcast (__main__.TestONNXUtils) ... ok
Jan 11 07:47:00 test_prepare_onnx_paddings (__main__.TestONNXUtils) ... ok
Jan 11 07:47:00 
Jan 11 07:47:00 ======================================================================
Jan 11 07:47:00 ERROR: setUpClass (__main__.TestHub)
Jan 11 07:47:00 ----------------------------------------------------------------------
Jan 11 07:47:00 Traceback (most recent call last):
Jan 11 07:47:00   File "/opt/conda/lib/python3.6/site-packages/urllib3/connection.py", line 159, in _new_conn
Jan 11 07:47:00     (self._dns_host, self.port), self.timeout, **extra_kw)
Jan 11 07:47:00   File "/opt/conda/lib/python3.6/site-packages/urllib3/util/connection.py", line 57, in create_connection
Jan 11 07:47:00     for res in socket.getaddrinfo(host, port, family, socket.SOCK_STREAM):
Jan 11 07:47:00   File "/opt/conda/lib/python3.6/socket.py", line 745, in getaddrinfo
Jan 11 07:47:00     for res in _socket.getaddrinfo(host, port, family, type, proto, flags):
Jan 11 07:47:00 socket.gaierror: [Errno -3] Temporary failure in name resolution
Jan 11 07:47:00 
Jan 11 07:47:00 During handling of the above exception, another exception occurred:
Jan 11 07:47:00 
Jan 11 07:47:00 Traceback (most recent call last):
Jan 11 07:47:00   File "/opt/conda/lib/python3.6/site-packages/urllib3/connectionpool.py", line 600, in urlopen
Jan 11 07:47:00     chunked=chunked)
Jan 11 07:47:00   File "/opt/conda/lib/python3.6/site-packages/urllib3/connectionpool.py", line 343, in _make_request
Jan 11 07:47:00     self._validate_conn(conn)
Jan 11 07:47:00   File "/opt/conda/lib/python3.6/site-packages/urllib3/connectionpool.py", line 839, in _validate_conn
Jan 11 07:47:00     conn.connect()
Jan 11 07:47:00   File "/opt/conda/lib/python3.6/site-packages/urllib3/connection.py", line 301, in connect
Jan 11 07:47:00     conn = self._new_conn()
Jan 11 07:47:00   File "/opt/conda/lib/python3.6/site-packages/urllib3/connection.py", line 168, in _new_conn
Jan 11 07:47:00     self, "Failed to establish a new connection: %s" % e)
Jan 11 07:47:00 urllib3.exceptions.NewConnectionError: <urllib3.connection.VerifiedHTTPSConnection object at 0x7f67479b2dd8>: Failed to establish a new connection: [Errno -3] Temporary failure in name resolution
Jan 11 07:47:00 
Jan 11 07:47:00 During handling of the above exception, another exception occurred:
Jan 11 07:47:00 
Jan 11 07:47:00 Traceback (most recent call last):
Jan 11 07:47:00   File "/opt/conda/lib/python3.6/site-packages/requests/adapters.py", line 449, in send
Jan 11 07:47:00     timeout=timeout
Jan 11 07:47:00   File "/opt/conda/lib/python3.6/site-packages/urllib3/connectionpool.py", line 638, in urlopen
Jan 11 07:47:00     _stacktrace=sys.exc_info()[2])
Jan 11 07:47:00   File "/opt/conda/lib/python3.6/site-packages/urllib3/util/retry.py", line 398, in increment
Jan 11 07:47:00     raise MaxRetryError(_pool, url, error or ResponseError(cause))
Jan 11 07:47:00 urllib3.exceptions.MaxRetryError: HTTPSConnectionPool(host='download.pytorch.org', port=443): Max retries exceeded with url: /models/resnet18-5c106cde.pth (Caused by NewConnectionError('<urllib3.connection.VerifiedHTTPSConnection object at 0x7f67479b2dd8>: Failed to establish a new connection: [Errno -3] Temporary failure in name resolution',))
Jan 11 07:47:00 
Jan 11 07:47:00 During handling of the above exception, another exception occurred:
Jan 11 07:47:00 
Jan 11 07:47:00 Traceback (most recent call last):
Jan 11 07:47:00   File "test_utils.py", line 492, in setUpClass
Jan 11 07:47:00     cls.resnet18_pretrained = models.__dict__['resnet18'](pretrained=True).state_dict()
Jan 11 07:47:00   File "/var/lib/jenkins/.local/lib/python3.6/site-packages/torchvision/models/resnet.py", line 175, in resnet18
Jan 11 07:47:00     model.load_state_dict(model_zoo.load_url(model_urls['resnet18']))
Jan 11 07:47:00   File "/opt/conda/lib/python3.6/site-packages/torch/utils/model_zoo.py", line 66, in load_url
Jan 11 07:47:00     _download_url_to_file(url, cached_file, hash_prefix, progress=progress)
Jan 11 07:47:00   File "/opt/conda/lib/python3.6/site-packages/torch/utils/model_zoo.py", line 72, in _download_url_to_file
Jan 11 07:47:00     u = urlopen(url, stream=True)
Jan 11 07:47:00   File "/opt/conda/lib/python3.6/site-packages/requests/api.py", line 75, in get
Jan 11 07:47:00     return request('get', url, params=params, **kwargs)
Jan 11 07:47:00   File "/opt/conda/lib/python3.6/site-packages/requests/api.py", line 60, in request
Jan 11 07:47:00     return session.request(method=method, url=url, **kwargs)
Jan 11 07:47:00   File "/opt/conda/lib/python3.6/site-packages/requests/sessions.py", line 533, in request
Jan 11 07:47:00     resp = self.send(prep, **send_kwargs)
Jan 11 07:47:00   File "/opt/conda/lib/python3.6/site-packages/requests/sessions.py", line 646, in send
Jan 11 07:47:00     r = adapter.send(request, **kwargs)
Jan 11 07:47:00   File "/opt/conda/lib/python3.6/site-packages/requests/adapters.py", line 516, in send
Jan 11 07:47:00     raise ConnectionError(e, request=request)
Jan 11 07:47:00 requests.exceptions.ConnectionError: HTTPSConnectionPool(host='download.pytorch.org', port=443): Max retries exceeded with url: /models/resnet18-5c106cde.pth (Caused by NewConnectionError('<urllib3.connection.VerifiedHTTPSConnection object at 0x7f67479b2dd8>: Failed to establish a new connection: [Errno -3] Temporary failure in name resolution',))
Jan 11 07:47:00 
Jan 11 07:47:00 ----------------------------------------------------------------------
Jan 11 07:47:00 Ran 17 tests in 28.658s
Jan 11 07:47:00 
Jan 11 07:47:00 FAILED (errors=1, skipped=3)
Jan 11 07:47:00 Traceback (most recent call last):
Jan 11 07:47:00   File "test/run_test.py", line 455, in <module>
Jan 11 07:47:00     main()
Jan 11 07:47:00   File "test/run_test.py", line 447, in main
Jan 11 07:47:00     raise RuntimeError(message)
Jan 11 07:47:00 RuntimeError: test_utils failed!
Jan 11 07:47:00 + cleanup
Jan 11 07:47:00 + retcode=1
Jan 11 07:47:00 + set +x

**Comment by gujinghui** [comment]
unknown failure. Should be not related to this PR

> update-initramfs: Generating /boot/initrd.img-4.4.0-96-generic
Exception:
Traceback (most recent call last):
  File "/usr/lib/python2.7/dist-packages/pip/basecommand.py", line 122, in main
    status = self.run(options, args)
  File "/usr/lib/python2.7/dist-packages/pip/commands/install.py", line 278, in run
    requirement_set.prepare_files(finder, force_root_egg_info=self.bundle, bundle=self.bundle)
  File "/usr/lib/python2.7/dist-packages/pip/req.py", line 1198, in prepare_files
    do_download,
  File "/usr/lib/python2.7/dist-packages/pip/req.py", line 1376, in unpack_url
    self.session,
  File "/usr/lib/python2.7/dist-packages/pip/download.py", line 572, in unpack_http_url
    download_hash = _download_url(resp, link, temp_location)
  File "/usr/lib/python2.7/dist-packages/pip/download.py", line 433, in _download_url
    for chunk in resp_read(4096):
  File "/usr/lib/python2.7/dist-packages/pip/download.py", line 421, in resp_read
    chunk_size, decode_content=False):
  File "/usr/share/python-wheels/urllib3-1.7.1-py2.py3-none-any.whl/urllib3/response.py", line 225, in stream
    data = self.read(amt=amt, decode_content=decode_content)
  File "/usr/share/python-wheels/urllib3-1.7.1-py2.py3-none-any.whl/urllib3/response.py", line 174, in read
    data = self._fp.read(amt)
  File "/usr/lib/python2.7/httplib.py", line 602, in read
    s = self.fp.read(amt)
  File "/usr/lib/python2.7/socket.py", line 380, in read
    data = self._sock.recv(left)
  File "/usr/lib/python2.7/ssl.py", line 341, in recv
    return self.read(buflen)
  File "/usr/lib/python2.7/ssl.py", line 260, in read
    return self._sslobj.read(len)
error: [Errno 104] Connection reset by peer

Storing debug log for failure in /home/circleci/.pip/pip.log
Exited with code 2

**Comment by gujinghui** [comment]
@yinghai
rebased. pls help review.

**Comment by yinghai** [comment]
@BIT-silence could you take a look?

File: caffe2/ideep/operators/expand_squeeze_dims_op.cc [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+  USE_IDEEP_DEF_ALIASES();
+  USE_IDEEP_OPERATOR_FUNCTIONS();
+  using FALLBACK_OP = IDEEPFallbackOp<ExpandDimsOp<CPUContext>, SkipIndices<0>>;
+
+  IDEEPExpandDimsOp(const OperatorDef& operator_def, Workspace* ws)
+      : IDEEPOperator(operator_def, ws),
+        fallback_(operator_def, ws) {
+    dims_ = OperatorBase::GetRepeatedArgument<int>("dims");
+    auto originalSize = dims_.size();
+    CAFFE_ENFORCE(originalSize > 0, "Parameter `dims` must be provided.");
```
Comment (BIT-silence) [inline caffe2/ideep/operators/expand_squeeze_dims_op.cc]:
```suggestion
    CAFFE_ENFORCE_GT(originalSize, 0, "Parameter `dims` must be provided.");
```

File: caffe2/ideep/operators/expand_squeeze_dims_op.cc [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+    std::sort(dims_.begin(), dims_.end());
+    dims_.erase(std::unique(dims_.begin(), dims_.end()), dims_.end());
+    if (dims_.size() < originalSize) {
+      LOG(WARNING) << "Parameter `dims` has repeated dimensions.";
+    }
+    CAFFE_ENFORCE(dims_.front() >= 0, "Dimension ids must be non-negative.");
```
Comment (BIT-silence) [inline caffe2/ideep/operators/expand_squeeze_dims_op.cc]:
```suggestion
    CAFFE_ENFORCE_GE(dims_.front(), 0, "Dimension ids must be non-negative.");
```

File: caffe2/ideep/operators/expand_squeeze_dims_op.cc [hunk 4]
```diff
@@ -0,0 +1,129 @@
+#include <caffe2/ideep/ideep_utils.h>
```
Comment (BIT-silence) [inline caffe2/ideep/operators/expand_squeeze_dims_op.cc]:
```suggestion
#include "caffe2/operators/expand_squeeze_dims_op.h"

#include <vector>

#include "caffe2/ideep/ideep_utils.h"
#include "caffe2/ideep/operators/operator_fallback_ideep.h"
```
https://google.github.io/styleguide/cppguide.html#Names_and_Order_of_Includes

File: caffe2/ideep/operators/expand_squeeze_dims_op.cc [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+    auto originalSize = dims_.size();
+    CAFFE_ENFORCE(originalSize > 0, "Parameter `dims` must be provided.");
+    std::sort(dims_.begin(), dims_.end());
+    dims_.erase(std::unique(dims_.begin(), dims_.end()), dims_.end());
+    if (dims_.size() < originalSize) {
+      LOG(WARNING) << "Parameter `dims` has repeated dimensions.";
+    }
+    CAFFE_ENFORCE(dims_.front() >= 0, "Dimension ids must be non-negative.");
+  }
+  virtual ~IDEEPExpandDimsOp() {}
```
Comment (BIT-silence) [inline caffe2/ideep/operators/expand_squeeze_dims_op.cc]: Remove virtual here, the class is marked as final

File: caffe2/ideep/operators/expand_squeeze_dims_op.cc [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+      newDims.insert(newDims.begin() + dim, 1);
+    }
+
+    Y->reshape(newDims);
+    return true;
+  }
+
+ private:
+  vector<int> dims_;
+  FALLBACK_OP fallback_;
```
Comment (BIT-silence) [inline caffe2/ideep/operators/expand_squeeze_dims_op.cc]:
```suggestion
  const FALLBACK_OP fallback_;
```

File: caffe2/ideep/operators/expand_squeeze_dims_op.cc [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+    for (const auto dim : dims_) {
+      newDims.insert(newDims.begin() + dim, 1);
+    }
+
+    Y->reshape(newDims);
+    return true;
+  }
+
+ private:
+  vector<int> dims_;
```
Comment (BIT-silence) [inline caffe2/ideep/operators/expand_squeeze_dims_op.cc]:
```suggestion
  std::vector<int> dims_;
```

File: caffe2/ideep/operators/expand_squeeze_dims_op.cc [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+  USE_IDEEP_DEF_ALIASES();
+  USE_IDEEP_OPERATOR_FUNCTIONS();
+  using FALLBACK_OP = IDEEPFallbackOp<SqueezeOp<CPUContext>, SkipIndices<0>>;
+
+  IDEEPSqueezeOp(const OperatorDef& operator_def, Workspace* ws)
+      : IDEEPOperator(operator_def, ws),
+        fallback_(operator_def, ws) {
+    dims_ = OperatorBase::GetRepeatedArgument<int>("dims");
+    auto originalSize = dims_.size();
+    CAFFE_ENFORCE(originalSize > 0, "Parameter `dims` must be provided.");
```
Comment (BIT-silence) [inline caffe2/ideep/operators/expand_squeeze_dims_op.cc]:
```suggestion
    CAFFE_ENFORCE_GT(originalSize, 0, "Parameter `dims` must be provided.");
```

File: caffe2/ideep/operators/expand_squeeze_dims_op.cc [hunk 9]
```diff
@@ -0,0 +1,129 @@
...
+
+    std::sort(dims_.begin(), dims_.end());
+    dims_.erase(std::unique(dims_.begin(), dims_.end()), dims_.end());
+    if (dims_.size() < originalSize) {
+      LOG(WARNING) << "Parameter `dims` has repeated dimensions.";
+    }
+    CAFFE_ENFORCE(dims_.front() >= 0, "Dimension ids must be non-negative.");
```
Comment (BIT-silence) [inline caffe2/ideep/operators/expand_squeeze_dims_op.cc]:
```suggestion
    CAFFE_ENFORCE_GE(dims_.front(), 0, "Dimension ids must be non-negative.");
```

File: caffe2/ideep/operators/expand_squeeze_dims_op.cc [hunk 10]
```diff
@@ -0,0 +1,129 @@
...
+  }
+  virtual ~IDEEPSqueezeOp() {}
```
Comment (BIT-silence) [inline caffe2/ideep/operators/expand_squeeze_dims_op.cc]: Remove virtual here, the class is marked as final.

File: caffe2/ideep/operators/expand_squeeze_dims_op.cc [hunk 11]
```diff
@@ -0,0 +1,129 @@
...
+
+    CAFFE_ENFORCE_GT(
+        X.ndims(),
+        dims_.back(),
+        "Input needs at least ",
+        (dims_.back() + 1),
+        " dimensions.");
+    const auto& ideep_dims = X.get_dims();
+    vector<int64_t> dims(ideep_dims.begin(), ideep_dims.end());
+    const auto& new_dims = SqueezeOp<IDEEPContext>::ComputeDims(dims, dims_);
```
Comment (BIT-silence) [inline caffe2/ideep/operators/expand_squeeze_dims_op.cc]:
```suggestion
    const auto new_dims = SqueezeOp<IDEEPContext>::ComputeDims(dims, dims_);
```
Though we have lifetime extension, I'd suggest not to rely on it.
https://abseil.io/tips/107

File: caffe2/ideep/operators/expand_squeeze_dims_op.cc [hunk 12]
```diff
@@ -0,0 +1,129 @@
...
+      ideep::direct_copy::compute(X, *Y);
+    }
+
+    Y->reshape(new_dims_ideep);
+    return true;
+  }
+
+ private:
+  vector<int> dims_;
+  FALLBACK_OP fallback_;
```
Comment (BIT-silence) [inline caffe2/ideep/operators/expand_squeeze_dims_op.cc]:
```suggestion
  const FALLBACK_OP fallback_;
```

File: caffe2/ideep/operators/expand_squeeze_dims_op.cc [hunk 13]
```diff
@@ -0,0 +1,129 @@
...
+      // Copy if not inplace
+      ideep::direct_copy::compute(X, *Y);
+    }
+
+    Y->reshape(new_dims_ideep);
+    return true;
+  }
+
+ private:
+  vector<int> dims_;
```
Comment (BIT-silence) [inline caffe2/ideep/operators/expand_squeeze_dims_op.cc]:
```suggestion
  std::vector<int> dims_;
```

File: caffe2/ideep/operators/expand_squeeze_dims_op.cc [hunk 14]
```diff
@@ -0,0 +1,129 @@
...
+    auto* Y = Output(OUTPUT);
+
+    CAFFE_ENFORCE_GT(
+        X.ndims(),
+        dims_.back(),
+        "Input needs at least ",
+        (dims_.back() + 1),
+        " dimensions.");
+    const auto& ideep_dims = X.get_dims();
+    vector<int64_t> dims(ideep_dims.begin(), ideep_dims.end());
```
Comment (BIT-silence) [inline caffe2/ideep/operators/expand_squeeze_dims_op.cc]:
```suggestion
    std::vector<int64_t> dims(ideep_dims.begin(), ideep_dims.end());
```

**Comment by gujinghui** [comment]
@yinghai 
Refined code based-on suggestion.

**Comment by gujinghui** [comment]
@yinghai
Refined code based-on suggestion.

**Comment by yinghai** [comment]
Test failures. Rebase?

**Comment by gujinghui** [comment]
Looks like many cases are failed due to network issue these days ...

**Review by yinghai — APPROVED** [review]

**Review by yinghai — CHANGES_REQUESTED** [review]
Test failure: 
```
Flaky: Hypothesis test_squeeze_fallback(self=<caffe2.caffe2.python.ideep.expanddims_squeeze_op_test.ExpandDimsSqueezeTest testMethod=test_squeeze_fallback>, squeeze_dims=[3], inplace=True, gc=, dc=[, device_type: 5]) produces unreliable results: Falsified on the first call but did not on a subsequent one
  File "/usr/local/fbcode/gcc-5-glibc-2.23/lib/python2.7/unittest/case.py", line 329, in run
    testMethod()
  File "/mnt/btrfs/trunk-hg-fbcode-fbsource-40439157-1549437637/fbcode/buck-out/dev/gen/caffe2/caffe2/python/ideep/expanddims_squeeze_op_test-2.7#binary,link-tree/caffe2/caffe2/python/ideep/expanddims_squeeze_op_test.py", line 34, in test_squeeze_fallback
    squeeze_dims=st.lists(st.integers(0, 3), min_size=1, max_size=3),
  File "/mnt/btrfs/trunk-hg-fbcode-fbsource-40439157-1549437637/fbcode/buck-out/dev/gen/caffe2/caffe2/python/ideep/expanddims_squeeze_op_test-2.7#binary,link-tree/hypothesis/core.py", line 553, in wrapped_test
    print_example=True, is_final=True
  File "/mnt/btrfs/trunk-hg-fbcode-fbsource-40439157-1549437637/fbcode/buck-out/dev/gen/caffe2/caffe2/python/ideep/expanddims_squeeze_op_test-2.7#binary,link-tree/hypothesis/executors.py", line 58, in default_new_style_executor
    return function(data)
  File "/mnt/btrfs/trunk-hg-fbcode-fbsource-40439157-1549437637/fbcode/buck-out/dev/gen/caffe2/caffe2/python/ideep/expanddims_squeeze_op_test-2.7#binary,link-tree/hypothesis/core.py", line 111, in run
    return test(*args, **kwargs)
  File "/mnt/btrfs/trunk-hg-fbcode-fbsource-40439157-1549437637/fbcode/buck-out/dev/gen/caffe2/caffe2/python/ideep/expanddims_squeeze_op_test-2.7#binary,link-tree/hypothesis/core.py", line 60, in test_or_flaky
    ) % (test.__name__, text_repr,))

Overwriting already registered item for key CPU
Overwriting already registered item for key CUDA
Overwriting already registered item for key HIP
test_squeeze_fallback (caffe2.caffe2.python.ideep.expanddims_squeeze_op_test.ExpandDimsSqueezeTest) ... Trying example: test_squeeze_fallback(self=<caffe2.caffe2.python.ideep.expanddims_squeeze_op_test.ExpandDimsSqueezeTest testMethod=test_squeeze_fallback>, squeeze_dims=[3], inplace=True, gc=, dc=[, device_type: 5])
Traceback (most recent call last):
  File "/mnt/btrfs/trunk-hg-fbcode-fbsource-40439157-1549437637/fbcode/buck-out/dev/gen/caffe2/caffe2/python/ideep/expanddims_squeeze_op_test-2.7#binary,link-tree/hypothesis/core.py", line 446, in evaluate_test_data
    search_strategy, test,
  File "/mnt/btrfs/trunk-hg-fbcode-fbsource-40439157-1549437637/fbcode/buck-out/dev/gen/caffe2/caffe2/python/ideep/expanddims_squeeze_op_test-2.7#binary,link-tree/hypothesis/executors.py", line 58, in default_new_style_executor
    return function(data)
  File "/mnt/btrfs/trunk-hg-fbcode-fbsource-40439157-1549437637/fbcode/buck-out/dev/gen/caffe2/caffe2/python/ideep/expanddims_squeeze_op_test-2.7#binary,link-tree/hypothesis/core.py", line 111, in run
    return test(*args, **kwargs)
  File "/mnt/btrfs/trunk-hg-fbcode-fbsource-40439157-1549437637/fbcode/buck-out/dev/gen/caffe2/caffe2/python/ideep/expanddims_squeeze_op_test-2.7#binary,link-tree/caffe2/caffe2/python/ideep/expanddims_squeeze_op_test.py", line 53, in test_squeeze_fallback
    Y0 = workspace.FetchBlob('Y0')
  File "/mnt/btrfs/trunk-hg-fbcode-fbsource-40439157-1549437637/fbcode/buck-out/dev/gen/caffe2/caffe2/python/ideep/expanddims_squeeze_op_test-2.7#binary,link-tree/caffe2/python/workspace.py", line 356, in FetchBlob
    result = C.fetch_blob(StringifyBlobName(name))
RuntimeError: [enforce fail at pybind_state.cc:182] ws->HasBlob(name). Can't find blob: Y0
```

**Comment by gujinghui** [comment]
@yinghai 
Sorry. Bug in test case.
Fixed. Pls help review.

**Comment by yinghai** [comment]
`caffe2_py2_mkl_ubuntu16_04_test` failed.

**Review by yinghai — APPROVED** [review]
Thanks!

**Pull request closed without merging**